pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go

```diff
@@ -311,6 +312,18 @@ func (it *fakeContractReaderInterfaceTester) GetChainWriter(_ *testing.T) types.
 
 func (it *fakeContractReaderInterfaceTester) DirtyContracts() {}
 
+func (it *fakeContractReaderInterfaceTester) StartContractReader(t *testing.T) {
+	fake, ok := it.impl.(*fakeContractReader)
+	assert.True(t, ok)
+	require.NoError(t, fake.Start(context.Background()))
+}
+
+func (it *fakeContractReaderInterfaceTester) CloseContractReader(t *testing.T) {
+	fake, ok := it.impl.(*fakeContractReader)
+	assert.True(t, ok)
+	require.NoError(t, fake.Close())
+}
```

Comment:
I'm having trouble seeing why we need these helper methods. `impl` is already a `ContractReader`, so we don't need to cast it to a specific fake type:
```suggestion
func (it *fakeContractReaderInterfaceTester) StartContractReader(t *testing.T) {
	require.NoError(t, it.impl.Start(context.Background()))
}

func (it *fakeContractReaderInterfaceTester) CloseContractReader(t *testing.T) {
	require.NoError(t, it.impl.Close())
}
```
Additionally, there is already a `GetContractReader` method, so callers that are doing this:
```go
	tester.StartContractReader(t)
	defer tester.CloseContractReader(t)
```
Could instead do this:
```go
	tester.GetContractReader(t).Start(ctx)
	defer tester.GetContractReader(t).Close()
```
Or better yet:
```go
	servicetest.Run(t, tester.GetContractReader(t))
```

Reply:
@jmank88 @silaslenihan  I couldn't do it this way. Tests fail for the loop relay client. 

Pushed some changes with a new idea:
- Modify `testCases` and `Setup` to receive a `startCR` flag and control whether we should start CR or not
- With this, `Start` and `Close` will be handled by `Setup` with `Cleanup`

All tests are passing, including new ones where we test that the service does not start. Let me know your thoughts

Reply:
Could you please update the description? I'm having trouble understanding the fundamental problem we are trying to solve.

Reply:
Since the `BasicTester` interface is used more generally than for just chain reader, I don't think it makes sense to add a chain reader only argument to `BasicTester.Setup`.

Reply:
> We can't just call Start and Close directly on the return of GetContractReader to achieve this. This is because we also run relay client loop tests here. We are calling Start and Close on the client and not on the server as would be required.

> The two solutions I proposed were aimed at solving this problem, the one with the helper methods and now the one with the flag.

Something isn't adding up here. The modification that I proposed to your helper methods was intended to execute _exactly_ equivalently. It was not proposing a change in behavior - definitely not one that switched from client to server or vice versa. If the helper method solution was valid, then we should revisit it, but in the most simplified form.

Reply:
Pushed a new solution with some comments on changes. If do you think it's okay we can go with it. Else, I'll revisit the one with the helper methods